src/ClassDiscovery.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Http\Discovery;
+
+/**
+ * Registry that based find results on class existence
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+abstract class ClassDiscovery
+{
+    /**
+     * Cached class
+     *
+     * @var object
+     */
```

Comment:
Be aware, that in this case tha cache property will be the same for all child classes. You either have to add a cache property to all discovery objects or handle this case in the parent, like saving the cache in an array based on get_called_class.